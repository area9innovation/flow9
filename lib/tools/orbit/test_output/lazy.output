QStandardPaths: wrong permissions on runtime directory /run/user/1000/, 0755 instead of 0700
Flow compiler (3rd generation)

Processing 'orbit' on http server
done in 0.12s

tests/lazy.orb: tests/lazy.orb:7:1: Parse error
};
^
app->exec() == 1
