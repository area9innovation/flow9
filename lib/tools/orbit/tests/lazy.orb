// Lazy evaluation test

// Define a function that takes an ast parameter (won't be evaluated until explicitly done)
fn delayedEval(expr : ast) = 
  println("Evaluating the expression"),
  eval(expr);

// Define a function that would cause side effects when evaluated
fn sideEffect() = 
  println("Side effect executed!"),
  42;

// Test 1: Call with lazily evaluated expression
println("Test 1: Should NOT see side effect until eval");
let result1 = delayedEval(sideEffect());
println("Result: " + i2s(result1));

// Test 2: Demonstrate conditional evaluation
fn conditionalEval(condition : bool, trueExpr : ast, falseExpr : ast) =
  condition is (
    c => 
      println("Evaluating true expression"),
      eval(trueExpr)
    if c;
    c => 
      println("Evaluating false expression"),
      eval(falseExpr)
  );

println("\nTest 2: Conditional evaluation, should only see one side effect");
let result2 = conditionalEval(true, sideEffect(), sideEffect());
println("Result: " + i2s(result2));

// Test 3: Delayed arithmetic
fn lazyAdd(a : ast, b : ast) = 
  println("Adding numbers"),
  eval(a) + eval(b);

println("\nTest 3: Lazy arithmetic, evaluations in order");
let result3 = lazyAdd(
  println("Computing first value"),
  10,
  println("Computing second value"),
  32
);
println("Result: " + i2s(result3))