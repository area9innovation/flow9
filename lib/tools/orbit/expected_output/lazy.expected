Test 1: Should NOT see side effect until eval
Evaluating the expression
sideEffect()
Side effect executed!
Result: 42
Test 2: Conditional evaluation, should only see one side effect
Evaluating true expression
Side effect executed!
Result: 42
Test 3: Lazy arithmetic, evaluations in order
Adding numbers
Result: 42
Result: "Result: 42"
